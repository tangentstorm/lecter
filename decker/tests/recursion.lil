

on addrec x y do
	if x>0
		1+addrec[x-1 y]
	else
		y
	end
end

on addtail x y do
	if x>0
		addtail[x-1 y+1]
	else
		y
	end
end

print["non-tail recursion"]
print[addrec[3 5]]

print["tail recursion: shallow"]
print[addtail[3 5]]

print["deepest observed stack: %i" sys.workspace.depth]

print["tail recursion: deep"]
print[addtail[800 5]]

print["deepest observed stack: %i" sys.workspace.depth]

print["false tail-calls"]
# this is a minimized example from a more elaborate
# program posing problems. while loops create scopes,
# they must not be treated as a possible site for a tail call:
on poison do
	y:0 # if a tail call, this modifies the outer scope!
end
each y in range 3
	each x in list 1 poison[] end
	print[y]
end
# should print
# 0
# 1
# 2

on factrec n a do
	# somewhat overly verbose tail-recursive factorial,
	# demonstrating that ifelse interacts with TCE properly:
	if     n=0 a
	elseif n=1 factrec[n-1 n*a]
	elseif n=2 factrec[n-1 n*a]
	else       factrec[n-1 n*a] end
end

print[factrec[12 1]]
print["deepest observed stack: %i" sys.workspace.depth]
