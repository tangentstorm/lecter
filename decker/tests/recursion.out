non-tail recursion
8
tail recursion: shallow
8
deepest observed stack: 6
tail recursion: deep
805
deepest observed stack: 6
false tail-calls
0
1
2
479001600
deepest observed stack: 6
